activerecord/test/cases/adapters/postgresql/transaction_nested_test.rb

```diff
@@ -206,5 +220,17 @@ def with_warning_suppression
         ActiveRecord::Base.clear_active_connections!
         ActiveRecord::Base.connection.client_min_messages = log_level
       end
+
+      # Essentially any query that should cause the next nested transaction to be a savepoint transaction.
+      def make_parent_transaction_dirty
+        Bit.take
+      end
+
+      def assert_current_transaction_is_savepoint_transaction
+        current_transaction = Sample.connection.current_transaction
+        unless current_transaction.is_a?(ActiveRecord::ConnectionAdapters::SavepointTransaction)
+          flunk("current transaction is not a savepoint transaction")
+        end
```

Comment:
I was just curious: is there is a benefit to writing it this way instead of using `assert_kind_of`?  If `assert_kind_of` fails and dumps the transaction object, is the output obnoxiously large?

Reply:
Not much benefit, to be honest. I just felt like I really wanted to make a clear distinction between what test asserts and some setup-verification checks. I feel like I'm bad at explaining so let me provide a short example, let's consider following tests:
```ruby
test "premium plan users can access exclusive content" do
  user = users(:user_1)
  content = contents(:content_1)
  assert_predicate user.plan, :premium?
  assert_predicate content, :exclusive? 

  assert AccessCheck.for(user).can_access?(content)
end
```
I feel a little uncomfortable having 3 assertions when the test is intended to test only one thing. The minitest reporter also going to show `3 assertions` which is slightly misleading. What I really wanted to do is just to verify that my setup is correct and whatever happens with `user_1` or `content_1` is not going to change the test setup without being noticed. So I would prefer something like:

```ruby
test "premium plan users can access exclusive content" do
  user = users(:user_1)
  content = contents(:content_1)
  flunk("not a premium user") unless user.plan.premium?
  flunk("not an exclusive content") unless content.exclusive?

  assert AccessCheck.for(user).can_access?(content)
end
```
And I'm hoping that this makes a clear separation between test setup and test assertions.
However I agree that this is something I personally prefer and it doesn't seem to be widely used or used in Rails. And also this example is oversimplified, I guess I could have called fixtures like `:premium_user` and `:exclusive_content` to reduce chances of those fixtures behave differently. Those fixtures suppose to represent more complex real-world examples, like initialization of `SavepointTransaction` which is happening behind the scenes and it's better to be sure that we do it in the test setup, otherwise we will be testing different scenario

Reply:
You are right about the output, I just tried with a dummy test like:
```ruby
    test "is not a savepoint transaction" do
      Sample.transaction do
        assert_kind_of(ActiveRecord::ConnectionAdapters::SavepointTransaction, Sample.connection.current_transaction)
      end
    end
```

and the output is not the most actionable:
<img width="644" alt="image" src="https://user-images.githubusercontent.com/5512772/159546044-0b406f75-c9c4-45e8-8501-47578d06abd4.png">

I could have passed `message` argument to the `assert_kind_of` method but one more thing I'm slightly concerned about is that `assert_kind_of` is counted as an assertion, but semantically I don't want it to be an assertion, my intention was just to verify test setup and not making any assertion just yet

<img width="371" alt="image" src="https://user-images.githubusercontent.com/5512772/159546180-19b9c4bb-a981-4fed-92f6-b4500bdb3ddf.png">

I know that these reports are probably not that important but I'm working on a side project that utilizes these reports and I want to keep them as reliable as possible :)